[2단계 - 자동차 경주 리팩터링] 조이썬(이영수) 미션 제출합니다.

youngsu5582 opened the pull request:
안녕하세요 스티치! : )

Step 2 는 스스로가 일종의 퀄리티 에 도달한거 같아서 간단한 부분들에 대해서만 리팩토링 했습니다.
( 부족한 부분이나 , 고도화 할 부분이 있다면 좀 더 지적 해주고 공격해도 완전 좋습니다 ☺️☺️. )

대신 , 다시 한번 리팩토링을 하며 생각한 것들에 대해서 조금 글을 남기겠습니다!!

### Controller 의 역활

도메인 단위로 쪼개며 구현을 해보며 느낀점은
마음만 먹으면 , Domain 단위에서도 모든 로직을 다 수행이 가능하다는 생각이 들었습니다.

Cars 에서 performRace(PowerGenerator powerGenerator)
라는 코드를 통해 현재 Controller 에서 하는
```java
private void performRace(final List<Car> cars, Integer raceCount) {
    List<RaceProgress> raceProgresses = new ArrayList<>();
    for (int i = 0; i < raceCount; i++) {
        performEachRace(cars);
        raceProgresses.add(RaceProgress.from(cars));
    }

    outputView.printAllRaceProgress(raceProgresses);
}

private void performEachRace(final List<Car> cars) {
    cars.forEach(car -> car.race(createPower()));
}

private int createPower() {
    return randomNumberGenerator.generate();
}
```
해당 로직들을 전부 도메인 으로 넣을 수 있다고 생각합니다.
그리고 , PowerGenerator 를 주입해줌으로써
오히려 테스팅 할때도 더 의도적으로 테스팅이 가능하다는 장점 마저 생각했습니다.

```java
public List<RaceProgress> performRace(int raceCount, NumberGenerator powerGenerator) {
    return IntStream.range(0, raceCount)
                    .mapToObj(index -> performEachRace(powerGenerator)).toList();
}

private RaceProgress performEachRace(NumberGenerator powerGenerator) {
    value.forEach(car -> car.race(powerGenerator.generate()));
    return RaceProgress.from(value);
}
```

이럴때 컨트롤러는 단순히 도메인 의 메소드 호출 역활 기능만 한다고 생각이 되는데
1. 입/출력의 값을 수행
2. 도메인 생성 및 행위 호출 

이 두가지 기능만 한다고 생각 하고 도메인에서 행위에 대한 코드를 짜도 되는지 궁금합니다

#### 일급 컬렉션의 기능

위의 질문에서 나온 추가적인 질문으로

도메인들을 가지는 일급 컬렉션의 기능이 어디까지 허용이 되는지에 대해 궁금함을 가지게 됐습니다

```markdown
### 자동차

- 자동차는 이름을 가진다
  - 이름은 다섯 글자 이하여야 한다
  - 중복된 이름을 가질 수 없다
- 자동차는 움직일 수 있다
  - 움직이는 조건은 파워이다
  - 파워가 4 이상이면 이동한다
  - 파워가 4 미만이면 이동하지 않는다
  ```

기능 요구사항 작성에 이런식으로
도메인 자체에 대한 기능에 대해서 작성하고 ,
도메인들을 포함한 컬렉션에 대해서는 적지 않는 방식을 선택했습니다.

해당 방식이 비즈니스 로직은 포함하고
비 개발자 & 내용을 모르는 사람 들에게 이해를 돕고 요구 사항을 더 알기 쉽다고 생각하기 때문입니다.

이러한 요구 사항 방식으로 접근할 때
일급 컬렉션들이 어디까지 기능을 담당할지 정하는 기준에 대해서 궁금합니다!
( 위 말한거처럼 Cars 가 기능을 다 수행해도 괜찮을 지 )

( 별도로 , 해당 요구 사항 방식 처럼 접근이 맞는지에 대해서도 
  간단히  👍 , 👎 로 피드백 해주시면 감사하겠습니다! - 제가 너무 많이 묻는거 같아서 그렇습니다,,,! )

위에 내용 말고도
제가 구현한 코드 방향과 로직 들이 맞는지 , 놓치거나 아쉽거나 고도화 할 부분들이 있는지 등
피드백 해주시면 감사히 재 리팩토링 해서 다시 리뷰 신청하겠습니다!

감사합니다!! 스티치! 🙂🙇‍♂️

---

추가로 , var 구문은 그냥 남발 하거나 막 사용해도 상관이 없나요??
성능 이슈 , 가독성 이슈 등등 ...

youngsu5582 committed:
refactor : 코딩 컨벤션에 따른 멤버 변수 순서 변경

private static final ( 상수 ) 최상단 위치

[관련 피드백 링크](https://github.com/woowacourse/java-racingcar/pull/713#discussion_r1493648800)

youngsu5582 committed:
refactor : 불필요한 Wrapper 클래스 제거

Integer -> int 로 변환

[관련 피드백 링크](https://github.com/woowacourse/java-racingcar/pull/713#discussion_r1493649855)

youngsu5582 committed:
refactor : RandomNumberGenerator 다시 추가

- CarConstant 삭제 , 상수 RandomNumberGenerator 에서 담당
- GameController 가 RandomNumberGenerator 가짐

[관련 피드백 링크](https://github.com/woowacourse/java-racingcar/pull/713#discussion_r1493649855)

youngsu5582 committed:
refactor : RandomNumberGenerator 가 NumberGenerator 구현하게 분리

- 테스트에 용이하게 하기 위해 NumberGenerator interface 사용

youngsu5582 committed:
refactor : Controller 에서 담당한 기능 Cars Domain 으로 분리

- 경기 진행하는 performEachRace , performRace GameController -> Cars

youngsu5582 committed:
test(Cars) : 변경한 로직에 대한 테스트 코드 추가

- 테스트 코드를 위한 SequenceNumberGenerator 생성

lxxjn0 commented on src/main/java/controller/GameController.java:
```diff
@@ -2,32 +2,34 @@
     private final OutputView outputView;
 
-    public GameController(InputView inputView, OutputView outputView) {
+    private final RandomNumberGenerator randomNumberGenerator;
```
[의견] 제 개인적인 의견으로는 컨트롤러가 해당 필드를 가질 필요는 없다고 생각합니다. 테스트를 위해서 최대한 의존성을 Main까지 끌어냈다고 생각하기엔 컨트롤러는 테스트 대상이 아니기에 컨트롤러 내부에서 RandomNumberGenerator를 생성해도 문제가 안된다고 생각합니다. 그리고 꼭 Main까지 빼내고 싶다고 하더라고 startGame 메서드의 파라미터로도 넘겨줄 수 있는 방법이 있고요.

객체가 가지는 프로퍼티(필드)는 객체의 존재 이유를 나타내는 값입니다. 그렇기에 이 객체의 존재 이유를 이 필드가 나타낼 수 있는가에 대해서 고민하면서 코드를 작성하신다면 더 도움이 될 수 있을 것 같습니다.

youngsu5582 replied:
해당 내용에 대해서도 생각한게
인수 테스트나 Main 클래스를 통한 
여러 가지 케이스에 대한 테스트를 고려해서 주입하는게 맞다고 생각했는데
결국 GameController 가 생성할 때 주입 한다의 합리적인 이유는 되지 못할거 같습니다

특히 , startGame 할때 파라미터로 넘겨주는게 맞는거 같습니다!

하지만 , 해당 부분을 수정하려고 하면
```java
private void performRace(final Cars cars, Integer raceCount,NumberGenerator powerGenerator) {
}
```

이렇게 매개변수 로 세개 가 필연적으로 들어가게 되는 경우가 생깁니다.
최대 2개의 매개 변수로 구현을 하고 싶은데 포기해야 하는걸까요??

lxxjn0 replied:
메서드의 파라미터를 최대한 적게 만드는 것이 좋긴 한데, 현실적으로 2개 이하의 파라미터만을 가지는 메서드를 유지하기란 쉽지 않을 것 같아요. 그렇다면 2개가 넘는 경우 모든 메서드는 dto와 같은 객체를 필요로 하게 되는데 이는 오히려 복잡성을 증가시킬 수도 있고요.

적절한 범위 내에서 유지 가능한 정도의 파라미터로 메서드를 관리한다고 생각하시고 그 개수는 개발자 개인에 따라 다를 수 있을 것 같습니다. (저 같은 경우 실제 사내에서는 상상하기 힘든 개수의 파라미터 개수를 가지는 메서드도 많이 만들었던.. 것 같아요 ㅠㅡㅠ)

lxxjn0 replied:
메서드의 파라미터를 최대한 적게 만드는 것이 좋긴 한데, 현실적으로 2개 이하의 파라미터만을 가지는 메서드를 유지하기란 쉽지 않을 것 같아요. 그렇다면 2개가 넘는 경우 모든 메서드는 dto와 같은 객체를 필요로 하게 되는데 이는 오히려 복잡성을 증가시킬 수도 있고요.

적절한 범위 내에서 유지 가능한 정도의 파라미터로 메서드를 관리한다고 생각하시고 그 개수는 개발자 개인에 따라 다를 수 있을 것 같습니다. (저 같은 경우 실제 사내에서는 상상하기 힘든 개수의 파라미터 개수를 가지는 메서드도 많이 만들었던.. 것 같아요 ㅠㅡㅠ)

lxxjn0 commented on src/main/java/controller/GameController.java:
```diff
@@ -37,28 +39,13 @@ private Cars registerCars() {
-        }
-
+    private void performRace(final Cars cars, Integer raceCount) {
+        List<RaceProgress> raceProgresses = cars.performRace(raceCount, randomNumberGenerator);
```
[질문] GameController가 가지는 역할은 전체 자동차 게임 진행 과정을 표현한 코드로 보여집니다. 그렇다면 cars의 performRace라는 메서드를 조금 더 cars라는 객체의 역할에 맞게 바꿔보면 어떨까요? 

직접적인 답변을 드리기는 아쉬워서 돌려서 말하다보니 오히려 이해하시기가 어려울 수도 있을 것 같긴 합니다.. ㅠㅠ 추가적인 설명을 드리자면 각각의 클래스 또는 객체가 어떠한 역할과 책임을 가지는 지 생각해보고, 이 객체의 메서드는 어떤 역할까지 하는게 맞는지, 그 이상의 역할은 어디에 담당하게 두는게 맞는지와 같은 부분을 초점으로 잡고 생각해보시면 좋을 것 같아요.

youngsu5582 replied:
제가 생각하기에는 
performRace 가 단순히 , void 형태의 각 Car 들의 race 를 수행하는게 맞는거 같은데

스티치의 생각이 이게 맞을까요??

```java
private void performRace(final Cars cars, Integer raceCount) {
    NumberGenerator powerGenerator = new RandomNumberGenerator();
    List<RaceProgress> raceProgresses = IntStream.range(0, raceCount)
                                                 .mapToObj(value -> performEachRace(cars, powerGenerator))
                                                 .toList();
    outputView.printAllRaceProgress(raceProgresses);
}

private RaceProgress performEachRace(final Cars cars, NumberGenerator powerGenerator) {
    cars.performRace(powerGenerator);
    return cars.recordRaceProgress();
}
```

Cars 는 단순히 경기 진행 후 -> 경기에 대한 과정 기록
과 같은 로직을 생각하여

Cars 가 수행을 한 후 -> 스스로 자신이 List<RaceProgress> 반환하는게 아닌
Cars 가 race 를 진행하고 -> 다시 한번 경기 과정으로 변환하는 로직을 가진후
Controller 에서 해당 RaceProgress 들을 저장하여 출력하는 기능을 수행하는걸로 수정했습니다.

해당 내용이 맞는거 같은데 , 제가 생각한게 맞을까요??! ✨✨

lxxjn0 replied:
넵넵!! 제가 생각하신 방향으로 말씀해주신 것 같아요 :)

제 생각에는 cars는 자동차의 진행을 관리하는 역할까지가 cars의 책임이 될 것 같고 게임에서 입력받은 횟수는 게임의 로직을 관리하는 지금의 controller가 처리해야 한다고 생각해요. 생각해주신 부분이 제가 의도한 방향과 일치하는 것 같습니다 👍🏼

lxxjn0 commented on src/main/java/util/NumberGenerator.java:
```diff
@@ -0,0 +1,5 @@
+package util;
+
+public interface NumberGenerator {
```
랜덤한 값을 생성해야 하는 부분을 인터페이스로 추출 후 테스트가 용이하게 코드를 변경하였네요 👍🏼

lxxjn0 commented on src/main/java/util/RandomNumberGenerator.java:
```diff
@@ -0,0 +1,11 @@
+import java.util.Random;
+
+public class RandomNumberGenerator implements NumberGenerator{
+    private static final int RANDOM_NUMBER_UPPER_BOUND = 10;
```
적절한 매직넘버 처리 👍🏼

lxxjn0 commented on src/test/java/mock/SequenceNumberGenerator.java:
```diff
@@ -0,0 +1,15 @@
+
+import java.util.List;
+
+public class SequenceNumberGenerator implements NumberGenerator {
+    List<Integer> mockingList = List.of(4, 7, 9, 3, 1, 6, 6, 7, 8);
+    Integer index = 0;
+
+    @Override
+    public int generate() {
+        return mockingList.get(index++);
+    }
+}
```
[제안] 만약 테스트에서 좀 더 다양하고 유연하게 NumberGenerator를 사용하고 싶다면 생성자로 시퀀스를 전달받아도 좋을 것 같아요! 그렇다면 테스트 코드 단에서 바로 Generator가 어떠한 숫자를 어느 시점에 반환할 지 추측도 가능할 것 같아 코드 이해에도 도움이 될 것 같습니다 :)

youngsu5582 replied:
이게 맞는거 같습니다!
생각하니까 , 굳이 내장 값으로 초기화할 필요가 없겠네요!

lxxjn0 reviewed (changes requested):
조이썬, 2단계 피드백 반영하시느라 수고 많으셨습니다 👍🏼 

1단계에서 많은 부분이 수정되어서 2단계는 크게 수정사항이 없었던 것 같네요 :) 제가 간단한 부분들에 대해서, 그리고 고민해보면 좋을 부분에 대해서 추가로 피드백 남겨두었으니 해당 부분만 한번 읽어보시고 피드백 반영 부탁드릴게요.

---
### Controller의 역할

조이썬이 작성한 대로 입출력 호출 그리고 도메인 행위 호출까지가 컨트롤러가 가져야 하는 역할 범위라고 생각합니다. 그리고 실제 자동차 게임을 진행하는 게임 과정은 도메인 로직으로 빠져서 관리되어야 하는 것이 맞다고 생각하고요. 

그러나 콘솔 어플리케이션 특성 상 이를 완전히 분리하기란 너무 어렵고 오히려 복잡한 코드가 만들어질 수 있습니다. 그렇기에 어느 정도까지는 타협점으로 가져가면서 Controller를 사용할 수 밖에 없다고 생각합니다. 그래서 이번 미션에서 Controller를 게임의 진행 과정과 입출력까지만 담당하도록 가정하고 클래스를 작성한다면, 그 부분까지는 충분히 납득가능한 영역이라는 생각을 합니다 :)

### 일급 컬렉션의 기능

조이썬의 비개발자나 또는 처음 코드를 접한 사람을 위한 접근은 정말 좋은 방향이라고 생각합니다 👍🏼 저도 늘 제 코드를 처음보는 누군가를 가정하면서 최대한 비즈니스 로직이 드러나고 이해하기가 쉽도록 코드를 짜려고 노력하고요.

그러나 코드의 구현 방법이나 언어의 기능, 또는 인프라의 환경 등 다양한 이유로 인해 실제 자연어로 표현한 비즈니스 내용과 코드는 달리 표현될 수 있다고 생각합니다. 기능 요구사항은 클래스 명세가 아닙니다. 그렇기에 꼭 Cars라는 클래스가 무슨 역할을 하는지가 드러날 필요는 없다고 생각합니다. 이 코드들이 어떠한 기능을 구현하였는지가 드러난다면 충분하지 않나 싶습니다.

추가로 일급 컬렉션은 오히려 비즈니스 로직의 구현보다는 코드의 응집도를 위한 코드적인 관점에서의 도구에 가깝다고 생각합니다. 자동차 게임의 비즈니스에서는 '자동차들을 일괄로 관리하기 위한 그 어떠한 제안이나 제약'은 존재하지 않습니다. 단지 코드 상에서 자동차들은 동일한 맥락으로 동작을 하고 이를 뭉쳐서 관리한다면 응집도 있는 코드를 작성할 수 있기에 그 도구 중 하나인 일급 컬렉션을 사용하자는 관점이 더 적절한 것 아닌가 싶습니다.

따라서 일급 컬렉션을 꼭 비즈니스 적인 구현 관점에서 바라보실 필요는 없고 코드 구현에 대한 방법 중 하나로 보셔도 충분할 것 같습니다 :)

추가로 더 궁금한 부분은 언제든지 질문주셔도 괜찮으니 편하게 질문 남겨주셔도 괜찮습니다 👍🏼

youngsu5582 committed:
refactor : SequenceNumberGenerator 생성자에서 리스트 초기화 하게 이동

[관련 피드백 링크](https://github.com/woowacourse/java-racingcar/pull/782#discussion_r1494451016)

youngsu5582 committed:
refactor : Cars 가 단순 car 들의 race 를 실행하게 분리 & RaceProgress 를 만들어주는 기능으로 분리

- peformRace 통해 단순히 race 실행
- convertRaceProgress 통해 RaceProgress 로 변환 담당

[관련 피드백 링크](https://github.com/woowacourse/java-racingcar/pull/782#discussion_r1494446854)

youngsu5582 committed:
refactor : GameController 변수로 NumberGenerator 제거

- 해당 내용에 대해 GameController - performRace 에 TODO 작성

[관련 피드백 링크](https://github.com/woowacourse/java-racingcar/pull/782#discussion_r1494442082)

youngsu5582 commented:
스티치 두번째 리뷰도 너무 감사합니다!

특히 , `일급 컬렉션은 코드의 응집도를 위한 코드적인 관점에서의 도구` 라는 말이
제게 참 좋았던 거 같습니다.

일급 컬렉션이 단순히 비즈니스 로직을 표현하는 방안이지 ,
기능 요구사항 과 비즈니스 로직에 드러내는게 아님을 깨달을 수 있는 좋은 기회 였던 거 같습니다!

차후  , 미션들 에서도 기능 요구사항을 정의할 때
단순히 개발 언어 ( Java , C , Python ... ) 등에 의존하지 않는 요구사항을 작성하기 위해 노력할 거 같습니다.

첫 번째 리뷰어가 스티치라서
도메인 단위적 으로 많은걸 깨닫고 가는거 같아서 좋았습니다!! 🥹🥹

youngsu5582 committed:
test(Cars) : 누락된 테스트 코드 추가

- 자동차들 상태 포함 컬렉션으로 만드는 convertRaceProgress 대한 테스트 코드 구현

lxxjn0 reviewed (approved):
조이썬, 자동차 경주 게임을 진행하시느라 수고 많으셨습니다 👍🏼 

처음 작성하신 코드에 비해서 지금의 코드는 많은 부분을 고민하고 개선하셨다는 느낌이 느껴지네요. 이번 레벨에서 고민하신 부분들을 계속 가져가시면서 다음 미션들도 진행하시기를 희망합니다 ㅎㅎㅎ 미션 진행하시느라 수고하셨습니다 :)
